### PR TITLE
feat: add default resolvers 

### LINKED ISSUE #134: Add optional support for external $ref in CLI
### **I'm submitting a...**
  - [ ] bug report
  - [x] feature request

### What is the current behavior?

The default behavior of Spectral v2.0 is to operate only on the original description object for 95% of the rules, and the uses the `resolvedTarget` in a few rules which specifically request that resolved version of the object (like oasOpsParams). 

Spectral CLI uses [json-ref-parser](https://github.com/stoplightio/json-ref-resolver) to created that resolvedTarget, and by default json-ref-parser only processes internal `$ref`. This is a safe default, because $ref can be potentially used maliciously if you are working with description files you don't control.  

This is unexpected for people transitioning from Speccy, which did external by default. See #108.

### What is the expected behavior?

Maybe we could have a new CLI option for the `lint` command, maybe `--external`.

The main `Spectral` constructor takes an object with a `resolver` key, which defaults to a stock standard json-ref-parser instance. The lint command should take advantage of this and change the resolver based on CLI options. 

```
if (flags.external) {
    const spectral = new Spectral(resolver: <fancy-resolver-however-that-works>);
}
```

Or something like that.

Thoughts @tbarn @marbemac @MikeRalphson?

### PR DESCRIPTION
**Pull request by marbemac**
fixes #134 #171 

WIP - still needs tests and docs.

### Please check if the PR fulfills these requirements

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce?

Bug fix / kind of a feature?

### If this is a feature change, what is the new behavior?

Will add description to readme.

### Does this PR introduce a breaking change?

No.

### Other information

### CODE REVIEW

File: src/resolvers/http.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import { Resolver } from '@stoplight/json-ref-resolver';
+import unfetch from 'unfetch';
+
+export const httpReader = {
+  async read(ref: any) {
+    return unfetch(String(ref)).then(r => r.text);
```
Comment (P0lip) [inline src/resolvers/http.ts]: code-style: you could use `await` here.

File: src/spectral.ts [hunk 2]
```diff
@@ -23,13 +25,13 @@ export * from './types';
...
 
-  private resolver: Resolver;
-  constructor(opts?: IConstructorOpts) {
-    this.resolver = opts && opts.resolver ? opts.resolver : new Resolver();
+  constructor(opts: IConstructorOpts = {}) {
+    this._resolver = opts.resolver ? opts.resolver : new Resolver();
   }
 
-  public async run(target: IParsedResult | object | string): Promise<IRuleResult[]> {
+  public async run(target: IParsedResult | object | string, opts: IRunOpts = {}): Promise<IRuleResult[]> {
```
Comment (P0lip) [inline src/spectral.ts]: This looks a bit weird. I believe we should let resolver be configurable rather than providing its settings to `run` method.
Thoughts?

File: src/cli/commands/lint.ts [hunk 3]
```diff
@@ -124,8 +126,11 @@ async function lint(name: string, flags: any, command: Lint, rules?: RuleCollect
   if (flags.verbose) {
     command.log(`Linting ${name}`);
   }
-  const spec: IParserResult = await readParsable(name, flags.encoding);
-  const spectral = new Spectral();
+
+  const targetUri = resolve(name);
+
+  const spec: IParserResult = await readParsable(targetUri, flags.encoding);
+  const spectral = new Spectral({ resolver: httpAndFileResolver });
```
Comment (P0lip) [inline src/cli/commands/lint.ts]: One more thought.
I'd say the above might be quite a significant change. It may slow down CI since Spectral will start making HTTP requests. We should provide a way to opt-out - a CLI flag here for resolver would be more than welcome.
It's a different story, though, so it can stay as is for now.

File: src/spectral.ts (hunk 2 above)
Comment (marbemac) [reply src/spectral.ts]: The `authority` is not a global resolver option, since it kind of depends on the file being run, so that option won't really work.

File: src/cli/commands/lint.ts (hunk 3 above)
Comment (marbemac) [reply src/cli/commands/lint.ts]: It will only make http requests if there are http $refs in the spec, and if there are http $refs in the spec, we can't really validate the spec without resolving them.

File: src/resolvers/http.ts (hunk 1 above)
Comment (marbemac) [reply src/resolvers/http.ts]: Yeah or we can just remove the async from line 5 :).

File: src/cli/commands/lint.ts (hunk 3 above)
Comment (P0lip) [reply src/cli/commands/lint.ts]: I'm aware, but IMHO we should provide a way to opt-out.

File: src/resolvers/http.ts (hunk 1 above)
Comment (P0lip) [reply src/resolvers/http.ts]: Yeah, you are also missing a function call 
```suggestion
    return unfetch(String(ref)).then(r => r.text());
```
Will fix it.

File: src/cli/commands/lint.ts (hunk 3 above)
Comment (philsturgeon) [reply src/cli/commands/lint.ts]: What would that actually mean though, would it just fail for folks using the CLI if they have HTTP refs? I think then opt out is removing the HTTP refs from your spec right?

File: src/cli/commands/lint.ts [hunk 4]
```diff
@@ -163,12 +169,17 @@ async function lint(name: string, flags: any, command: Lint, rules?: RuleCollect
...
-      source: resolve(process.cwd(), name),
+      source: targetUri,
       parsed: spec,
       getLocationForJsonPath,
     };
 
-    results = await spectral.run(parsedResult);
+    results = await spectral.run(parsedResult, {
+      resolve: {
+        authority: targetUri,
```
Comment (philsturgeon) [inline src/cli/commands/lint.ts]: Why is it called authority? It's not quite clicking for me yet.

File: src/cli/commands/lint.ts (hunk 3 above)
Comment (P0lip) [reply src/cli/commands/lint.ts]: > What would that actually mean though, would it just fail for folks using the CLI if they have HTTP refs? I think then opt out is removing the HTTP refs from your spec right?

Opting-out would mean turning off http/https $ref resolution.
Resolving external resources may significantly slow down CI. Not a big deal though, we can leave it as is.

Comment (marbemac) [reply src/cli/commands/lint.ts]: Yeah opting out would mean it would fail for folks if they use http $refs. I can't see a use case for this - how about we re-consider if somebody brings it up later?

File: src/cli/commands/lint.ts (hunk 4 above)
Comment (marbemac) [reply src/cli/commands/lint.ts]: Authority is just piggybacking off of the `authority` option in https://github.com/stoplightio/json-ref-resolver. I created json-ref-resolver quite a while ago, and in hindsight it's not the best name for that property. I have an open ticket to make an update to json-ref-resolver to use https://github.com/microsoft/vscode-uri rather than urijs (to fix a bunch of issues with windows style file paths), and was thinking about leveraging that breaking release to also update some of the names.

We could decide on a name right now, and use that in spectral so that we don't have a breaking change here later? The authority option of json-ref-resolver is described here: https://github.com/stoplightio/json-ref-resolver#example-resolving-relative-remote-references-with-the-authority-option.

Basically, it's the location of the document currently being resolved, so that relative $refs inside of it can be effectively resolved (solution to #171).

Open to suggestions, here are some:

- parentUri
- documentUri
- uri
- $id (from json schema, "It declares a base URI against which $ref URIs are resolved."). This is basically exactly how we're using it.
- documentId

File: package.json [hunk 5]
```diff
@@ -71,7 +72,8 @@
     "node-fetch": "2.x.x",
     "strip-ansi": "^5.2.0",
     "text-table": "^0.2.0",
-    "typescript-json-schema": "^0.38.0"
+    "typescript-json-schema": "^0.38.0",
+    "unfetch": "^4.1.0"
```
Comment (philsturgeon) [inline package.json]: We've been using node-fetch which uses node Request and other core stuff internally, meaning we can use tools like Polly.js for testing. Would you mind switching to that?

File: src/cli/commands/lint.ts (hunk 3 above)
Comment (philsturgeon) [reply src/cli/commands/lint.ts]: Yeah we've had a lot of complaints that it cant resolve and I cant imagine it going the other way. I agree we should reconsider if somebody asks for a way to turn it off. Maybe for security reasons or something else.

File: src/cli/commands/lint.ts (hunk 4 above)
Comment (philsturgeon) [reply src/cli/commands/lint.ts]: OHHH i see! documentUri sounds great. Maybe we could switch to vscode-uri and do all the breaky things in one go?

Comment (marbemac) [reply src/cli/commands/lint.ts]: > I have an open ticket to make an update to json-ref-resolver to use https://github.com/microsoft/vscode-uri rather than urijs (to fix a bunch of issues with windows style file paths), and was thinking about leveraging that breaking release to also update some of the names.

:)

File: package.json (hunk 5 above)
Comment (marbemac) [reply package.json]: Sure, good idea!

**Comment by marbemac** [comment]
Alright we have a tricky question...

Should we report errors in $ref'd files? For example, in the below image the `library.yaml` file has an error (the example does not match the schema). `one.yaml` references the address model in `library.yaml`. If a user runs spectral on `library.yaml`, should the error in `one.yaml` be reported? If so, it brings up the tricky issue of how to report this error to the end user in a way that is not confusing (e.g. it should probably say its in a different file, with accurate line numbers relevant to that file, etc). This is non-trivial to do.

Screen Shot 2019-06-04 at 8 08 36 PM

**Comment by P0lip** [comment]
@marbemac I think I've already asked you that question before and you said we wouldn't report such errors. Only errors relevant to linted files would be caught.

File: src/resolvers/http-and-file.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            if (err) reject(err);
+            resolve(data);
+          });
+        });
+      },
+    },
+  },
+
+  parseAuthorityResult: async opts => {
+    const parts = opts.targetAuthority.path().split('.');
```
Comment (P0lip) [inline src/resolvers/http-and-file.ts]: Using path.extname is likely to be a better idea here.

Comment (philsturgeon) [reply src/resolvers/http-and-file.ts]: A yeah I can see this causing trouble.

**Comment by marbemac** [comment]
Ok ready for another review. In answer to my question above, I think by default we just validate the provided file (not files that this file references). At a later date we could add an option to also validate $ref'd documents (would be off by default, so non breaking) - it will be tricky to get this UX right.

**Comment by philsturgeon** [comment]
This didn’t update the README and after this discussion I’m not sure what actually changed. If we are not validating the contents of external refs then what are we doing?

**Comment by philsturgeon** [comment]
@P0lip hey this still had "WIP - still needs tests and docs." on it, so we dont have tests, or docs. Could you maybe switch to doing this and put a pause on request maker?

**Comment by P0lip** [comment]
@philsturgeon we wanted to have browser testing first, didn't we?

**Comment by marbemac** [comment]
> This didn’t update the README and after this discussion I’m not sure what actually changed. If we are not validating the contents of external refs then what are we doing?

We're resolving file and http(s) $refs now, and reporting an error if they cannot be resolved. See this comment for the "do we validate things that this file references" question: https://github.com/stoplightio/spectral/pull/245#issuecomment-498899833

In general, that is a larger task and should probably be handled separately. It kind of changes an underlying assumption that we're validating one file at a time. We'll need to re-think how we track and report errors, since now errors (and their line numbers etc) can span across multiple files in a single spectral run.

cc @philsturgeon

File: src/cli/commands/lint.ts [hunk 7]
```diff
@@ -137,8 +140,15 @@ async function lint(name: string, flags: any, command: Lint, rules?: RuleCollect
   if (flags.verbose) {
     command.log(`Linting ${name}`);
   }
-  const spec: IParserResult = await readParsable(name, flags.encoding);
-  const spectral = new Spectral();
+
+  let targetUri = name;
+  if (!startsWith(name, 'http')) {
```
Comment (philsturgeon) [inline src/cli/commands/lint.ts]: @P0lip looks like if name is `"httpModels/somefile.json" it will think it is a URL. Can we switch to`http(s)://` regex in the test PR you are working on? #264

**Comment by philsturgeon** [comment]
Ok we've made a CLI specific issue for this. Thank you!

File: src/cli/commands/lint.ts (hunk 7 above)
Comment (P0lip) [reply src/cli/commands/lint.ts]: @philsturgeon sure, I'll make that happen.
